openedx/adg/lms/applications/tests/test_admin/test_user_application_adg_admin.py

```diff
@@ -0,0 +1,635 @@
+
+
+@pytest.mark.parametrize(
+    'request_method, status, note', [
+        ('POST', None, NOTE),
+        ('POST', UserApplication.WAITLIST, NOTE),
+        ('GET', None, '')
+    ],
+    ids=['post_without_status', 'post_with_status', 'get']
+)
+@pytest.mark.django_db
+@mock.patch('openedx.adg.lms.applications.admin.admin.ModelAdmin.changeform_view')
+@mock.patch('openedx.adg.lms.applications.admin.get_extra_context_for_application_review_page')
+@mock.patch('openedx.adg.lms.applications.admin.UserApplicationADGAdmin._save_application_review_info')
+def test_changeform_view(
+    mock_save_application_review_info,
+    mock_get_extra_context_for_application_review_page,
+    mock_changeform_view,
+    request,
+    user_application,
+    user_application_adg_admin_instance,
+    request_method,
+    status,
+    note
+):
+    """
+    Test the overridden changeform_view.
+
+    Test Case 1: post_without_status
+        Test that if a POST request is made with an internal note but without status, the application should not be
+        saved and admin should be shown the application review page with the right context.
+
+    Test Case 2: post_with_status
+        Test that if a POST request is made with internal note and status, the application should be updated and saved.
+
+    Test Case 3: get
+        Test that if a GET request is made, the right context should be rendered for the application review page.
+    """
+    application_id = user_application.id
+
+    expected_context = {'test_key': 'test_value'}
+    mock_get_extra_context_for_application_review_page.return_value = expected_context
+
+    request.method = request_method
+    if request.method == 'POST':
+        request.POST = {'internal_note': note}
+        if status:
+            request.POST['status'] = status
+            UserApplicationADGAdmin.changeform_view(user_application_adg_admin_instance, request, application_id)
+
+            mock_save_application_review_info.assert_called_once_with(user_application, request, note)
+            mock_changeform_view.assert_called_once_with(request, application_id, extra_context=None)
+        else:
+            UserApplicationADGAdmin.changeform_view(user_application_adg_admin_instance, request, application_id)
+
+            mock_save_application_review_info.assert_not_called()
+            mock_get_extra_context_for_application_review_page.assert_called_once_with(user_application)
+            mock_changeform_view.assert_called_once_with(request, application_id, extra_context=expected_context)
+    elif request.method == 'GET':
+        UserApplicationADGAdmin.changeform_view(user_application_adg_admin_instance, request, application_id)
+
+        mock_get_extra_context_for_application_review_page.assert_called_once_with(user_application)
+        mock_changeform_view.assert_called_once_with(request, application_id, extra_context=expected_context)
```

Comment:
In this particular case, for each set of parameter we have different set of assertion and we have implemented different conditions to separate out different cases. So it is not a good case to handle with parametrization. Let's please write separate test cases for each case, which would be comparatively more straightforward, easy to understand and fix, with minimal possible duplication.